.Net MEVD: LINQ-based filtering

roji opened the pull request:
## LINQ-based filtering

This implements LINQ-based filtering for all current implementations of IVectorStore (with the exception of Pinecone, blocking on #10415). This implements support for:

* Equality and inequality
* Comparison (GreaterThan, GreaterThanOrEqual, LessThan, LessThanOrEqual)
* Logical operators (and, or, not)
* Contains over collection field (`r => r.Tags.Contains("foo")`)
* Contains over inline collection (`r => new[] { "foo", "bar" }.Contains(r.String)`)

Notably missing at the moment is support for VectorStoreGenericDataModel in LINQ querying; I've opened #10468 and will do this a bit later, with additional planned work on the generic data model.

## API naming

Vector databases consistency refer to this feature as filter/filtering; in some cases (e.g. Chroma), `where` is used, following SQL nomenclature:

* [Qdrant](https://qdrant.tech/documentation/beginner-tutorials/search-beginners/#narrow-down-the-query)
* [Azure AI Search](https://learn.microsoft.com/en-us/azure/search/vector-search-overview#what-scenarios-can-vector-search-support)
* [Weaviate](https://weaviate.io/developers/weaviate/search/similarity#filter-results)
* [Milvus](https://milvus.io/docs/filtered-search.md)
* [Pinecone](https://docs.pinecone.io/guides/data/query-data#external-embedding)
* [Chroma](https://docs.trychroma.com/docs/querying-collections/metadata-filtering) (the term `where` is used in the API, but filtering is also used in the docs)

It seems the choice of exposing this via a `VectorSearchOptions.Filter` is indeed the correct one. Since we want to maintain backwards compatibility for existing users and preserve the current legacy filtering mechanism, this PR must introduce a separate, temporary `NewFilter` for the new LINQ-based filteirng. When releasing the stable version, we'll need to remove the old `Filter` property rename `NewFilter` to `Filter` (breaking users which switched to `NewFilter`).

## Custom mapping

The samples currently contain a MappingVectorStoreRecordCollection, which is a decorator transforming a public, user-facing record type to another internal one - the user provides two delegates to handle the conversion from the public record type to the internal, and vice versa. Such conversions become considerably more complex, since the user provides an expression tree referencing public record properties.

To make this work, the user would need to provide some logic to translate property accesses on the public record type to the internal one. For flat records, this could be a simple API where the user simply implements some interface translating public to internal; but for hierarchical records this gets a bit more complicated. I'd suggest deferring such advanced, arbitrary mapping scenarios to post-release. In the meantime, I've commented out MappingVectorStoreRecordCollection (and also Step4_NonStringKey_VectorStore) from the samples.

Note that the problem here doesn't come from LINQ per se, but from the fact that the user specifies **model** property names (the legacy search API had them specify **storage** property names instead, side-stepping the problem but causing an inconsistent API where sometimes model properties are used, and sometimes storage properties).

## Integration test suite

In addition, this implements new integration test projects for all IVectorStore connectors, implementing most of https://github.com/microsoft/semantic-kernel/issues/10194#issuecomment-2640276163:

* A shared FiltersTestBase class defines all LINQ filter tests, and is extended and used by all providers ([link](https://github.com/microsoft/semantic-kernel/pull/10273/files#diff-54e49edd3099d4031761dc5017c0e6bf491e6b334de33f300a54fa1ac3269ee6)). This can serve as the basic "spec" for what filters are supported across connector (subject to database support, of course).
* Each connector has its own separate project
* Testcontainers are used for Mongo, PostgreSQL, Qdrant, Redis, Sqlite, and Weaviate. Azure AI Search, CosmosNoSQL and CosmosMongoDB require configuration (environment variables, settings.json) to provide endpoint information, and tests are skipped if that's not provided.

/cc @westey-m @dmytrostruk 

Closes https://github.com/microsoft/semantic-kernel/issues/10156
Does most of https://github.com/microsoft/semantic-kernel/issues/10194

roji commented on dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantFilterTranslator.cs:
```diff
@@ -0,0 +1,178 @@
+        var right = this.Visit(andAlso.Right);
+
+        // Qdrant doesn't allow arbitrary nesting of logical operators, only one MUST list (AND), one SHOULD list (OR), and one MUST_NOT list (AND NOT).
+        // We can combine MUST and MUST_NOT; but we can only combine SHOULD if it's the *only* thing on the one side (no MUST/MUST_NOT), and there's no SHOULD on the other (only MUST/MUST_NOT).
```
Am not 100% sure about Qdrant's capabilities here, should confirm.

roji replied:
Fully implement all logical operators.

roji commented on dotnet/src/Connectors/VectorData.Abstractions/VectorSearch/IVectorizableTextSearch.cs:
```diff
@@ -20,6 +20,6 @@ public interface IVectorizableTextSearch<TRecord>
     Task<VectorSearchResults<TRecord>> VectorizableTextSearchAsync(
         string searchText,
-        VectorSearchOptions? options = default,
+        VectorSearchOptions<TRecord>? options = default,
```
Making VectorSearchOptions generic over TRecord isn't trivial... Modern C# users can use target-typed new, and in that case this doesn't interfere in the method calling. But people not doing target-typed new would have to explicitly specify the full type every time (even if they don't use filtering at all), which maybe isn't ideal. The alternative would be to move the LINQ filter to an optional parameter directly on the method (no strong feelings from my side here).

westey-m replied:
Since we are thinking of also making property selection an expression, that may result in a few more optional parameters on the method signature, to avoid making options generic.  For hybrid search it would be 3 additional, which isn't great, especially considering that 2 of those would very rarely be used, i.e. only if there is more than one dense vector or more than one fulltextsearch property on the record.

dmytrostruk commented on dotnet/SK-dotnet.sln:
```diff
@@ -439,6 +439,12 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "sk-chatgpt-azure-function",
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "kernel-functions-generator", "samples\Demos\CreateChatGptPlugin\MathPlugin\kernel-functions-generator\kernel-functions-generator.csproj", "{78785CB1-66CF-4895-D7E5-A440DD84BE86}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "VectorDataTests", "src\VectorDataTests", "{4F381919-F1BE-47D8-8558-3187ED04A84F}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "QdrantTests", "src\VectorDataTests\QdrantTests\QdrantTests.csproj", "{27D33AB3-4DFF-48BC-8D76-FB2CDF90B707}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "VectorDataSpecificationTests", "src\VectorDataTests\VectorDataSpecificationTests\VectorDataSpecificationTests.csproj", "{B29A972F-A774-4140-AECF-6B577C476627}"
```
With these project/folder names it's not clear whether these are integration or unit tests. I think it's important because sometimes there could be 2 test classes with the same name in a codebase, and when you search it - project/folder name allows you to distinguish them. So, if we can do `VectorDataIntegrationTests` or `QdrantIntegrationTests` and `VectorDataSpecificationIntegrationTests`, I think it should be helpful.

roji replied:
Sure, we can. I don't think it makes much sense to have unit tests for these providers (whose main thing is to communicate with their respective databases), but I can rename.

dmytrostruk commented on dotnet/src/Connectors/VectorData.Abstractions/VectorSearch/VectorSearchOptions.cs:
```diff
@@ -1,17 +1,26 @@
     /// Gets or sets a search filter to use before doing the vector search.
     /// </summary>
+    [Obsolete("Use NewFilter")] // TODO: Remove
     public VectorSearchFilter? Filter { get; init; }
```
Should we mark `VectorSearchFilter` as obsolete as well?

dmytrostruk commented on dotnet/src/VectorDataTests/VectorDataSpecificationTests/Filter/BasicFilterTestsBase.cs:
```diff
@@ -0,0 +1,96 @@
+    where TKey : notnull
+{
+    [Fact]
+    public virtual async Task Equality_with_int()
```
If we can keep PascalCase as naming convention for test methods like in other tests in a codebase - that would be great :)

dmytrostruk commented on dotnet/src/VectorDataTests/QdrantTests/Support/TestContainer/QdrantBuilder.cs:
```diff
@@ -0,0 +1,54 @@
+using Docker.DotNet.Models;
```
```suggestion
﻿// Copyright (c) Microsoft. All rights reserved.

using Docker.DotNet.Models;
```

dmytrostruk commented on dotnet/src/VectorDataTests/QdrantTests/Support/TestContainer/QdrantConfiguration.cs:
```diff
@@ -0,0 +1,51 @@
+using Docker.DotNet.Models;
```
```suggestion
﻿// Copyright (c) Microsoft. All rights reserved.

using Docker.DotNet.Models;
```

dmytrostruk commented on dotnet/src/VectorDataTests/QdrantTests/Support/TestContainer/QdrantContainer.cs:
```diff
@@ -0,0 +1,6 @@
+using DotNet.Testcontainers.Containers;
```
```suggestion
﻿// Copyright (c) Microsoft. All rights reserved.

using DotNet.Testcontainers.Containers;
```

westey-m commented:
@roji, a good scenario to also include in early exploration would be tag based operations, e.g. where there is an array of strings (tags) in a field in the database, and we want to check if one or more out of tags we provide match those.
We have received consistent feedback that this is a popular scenario for vector search filtering.

westey-m commented on dotnet/src/VectorDataTests/VectorDataSpecificationTests/Filter/BasicFilterTestsBase.cs:
```diff
@@ -0,0 +1,69 @@
+            e.Int2 == a.Int2);
+    }
+
+    [Fact]
```
One of the issues I have been facing with abstract base class tests is that a few of the DBs don't perform well on the Build server for integration tests, so we typically run them manually. Mostly issues with failed startup and flakiness.
I generalized some work done for Pinecone that allows us to have attribute based test disabling, which allows us to have an attribute on the inheriting class that disables all tests in the inheriting and base class.
Check out `VectorStoreFact`, `PineconeApiKeySetConditionAttribute` and `DisableVectorStoreTestsAttribute` for how this works.
Passing a Skip reason to Fact doesn't work because it requires a static reason, and it would just disable the tests for all inheritors.

roji replied:
Yeah, I remember this... I'll take a look.

But just off the top of my head, in CI we could run tests only for those providers where there's a good container-baesd option (like Qdrant), and not run them for the others (splitting the test projects by provider could also help with that a bit). Locally, developers can simply explicitly choose what it is they want to run - though that would mean that running all tests in the solution would produce failures. If that's not acceptable, a conditional [Fact] attribute like what you mention above could be good, based on some environment variable or whatever.

In any case, my hope is that this could be orthogonal to the tests themselves being general across providers, as opposed to written per-provider... But we'll see.

westey-m replied:
I quite like the idea of the test conditionality being based on whether the required secrets have been configured like in PineconeApiKeySetConditionAttribute, since it removes the need to change build settings in addition to setting the build secrets. Locally, I prefer skipped to failure, since failures need to be investigated, while the meaning of something being skipped is clear without further investigation.

roji replied:
Makes sense, will do that.

roji commented:
> a good scenario to also include in early exploration would be tag based operations, e.g. where there is an array of strings (tags) in a field in the database, and we want to check if one or more out of tags we provide match those.
We have received consistent feedback that this is a popular scenario for vector search filtering.

@westey-m thanks - will do that, it should be easily representible in LINQ by doing `b => b.Tags.Contains("foo")`. Maybe let's talk tomorrow about how that's represented e.g. in Qdrant (just because I'm currently focusing on it).

roji commented:
Note: there are some baseline issues failing the build (presumably because of the public API changes?). I'm not familiar with the mechanism used here, maybe someone could lend a hand?

dmytrostruk commented:
> Note: there are some baseline issues failing the build (presumably because of the public API changes?). I'm not familiar with the mechanism used here, maybe someone could lend a hand?

@roji I fixed the issue with public API changes. FYI to generate suppression files, you can run `dotnet msbuild SK-dotnet.sln -p:Configuration=Release -p:ApiCompatGenerateSuppressionFile=true` and it should fix the issue. 

Although there is another issue now: 
`error CS1685: The predefined type 'RuntimeHelpers' is defined in multiple assemblies in the global alias; using definition from 'Microsoft.SemanticKernel.Abstractions, Version=1.34.0.0, Culture=neutral, PublicKeyToken=null' [/home/runner/work/semantic-kernel/semantic-kernel/dotnet/src/SemanticKernel.Core/SemanticKernel.Core.csproj::TargetFramework=netstandard2.0]`

It looks like it's related to `IndexRange.cs` file that is added in this PR.

RogerBarreto commented on dotnet/samples/GettingStartedWithVectorStores/Step4_NonStringKey_VectorStore.cs:
```diff
@@ -1,195 +1,198 @@
-        }
-    }
-}
+// TODO: See note in MappingVectorStoreRecordCollection
```
Should we just delete it or add Obsolete?

roji replied:
Note that this isn't part of the public API - it's just a sample - so I'm not sure obsoleting makes sense. Also, obsoletion is generally for something that still works but is no longer the recommended way, whereas this no longer works; or rather, it can be made to work for the legacy/obsolete filter, but not (easily) for the new one.

@westey-m any opinion on what to do here? We can also leave commented out for now and deal with it later.

RogerBarreto commented on dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSqlFilterTranslator.cs:
```diff
@@ -0,0 +1,284 @@
+
+namespace Microsoft.SemanticKernel.Connectors.AzureCosmosDBNoSQL;
+
+internal class AzureCosmosDBNoSqlFilterTranslator
```
Is it possible to share some commom implementation between the translators (InternalUtilities) or are they very specific in all aspects? We might consider using a `core` utility for common behaviors.

roji replied:
There are some limited opportunities for reuse here, I think; I do think it may make sense to have a preprocessing pass on the incoming expression tree performed by a general component, which would (a) inline captured variables (for most databases), (b) normalize the different variants of Contains to Enumerable.Contains; that would allow the per-connector translators to be a bit simpler. But that doesn't go very far, and I'd rather do that in a future separate PR.

BTW I'm generally trying to move us away from shared internal code a bit, since that doesn't work for connectors that are implemented outside the repo; ideally, if something really makes sense as cross-connector support, it should be made a public, connector-facing API in MEVD itself.

RogerBarreto commented on dotnet/src/Connectors/Connectors.Memory.Postgres/PostgresVectorStoreDbClient.cs:
```diff
@@ -172,21 +174,23 @@ public async Task DeleteAsync<TKey>(string tableName, string keyColumn, TKey key
-    public async IAsyncEnumerable<(Dictionary<string, object?> Row, double Distance)> GetNearestMatchesAsync(
-        string tableName, IReadOnlyList<VectorStoreRecordProperty> properties, VectorStoreRecordVectorProperty vectorProperty, Vector vectorValue, int limit,
-        VectorSearchFilter? filter = default, int? skip = default, bool includeVectors = false, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+#pragma warning disable CS0618 // VectorSearchFilter is obsolete
```
I wonder if there's another non obsolete API that could be leveraged instead of ignoring the obsolescence, or keeping a tracking issue to tackle this later on.

roji replied:
The obsolescence here is specifically for the existing (legacy) filter, which we decided to preserve for now, to avoid breaking existing users - the plan is to remove that API (and rename NewFilter to Filter) before we actually release. So all these suppressions will go away in the next 2-3 months.

westey-m commented on dotnet/samples/GettingStartedWithVectorStores/Step2_Vector_Search.cs:
```diff
@@ -53,6 +53,7 @@ internal static async Task<VectorSearchResult<Glossary>> SearchVectorStoreAsync(
         return searchResultItems.First();
     }
 
+#pragma warning disable CS0618 // VectorSearchFilter is obsolete
```
Can we not switch to the new filter here?

roji replied:
Good point, will do.

westey-m commented on dotnet/src/SemanticKernel.Abstractions/Data/TextSearch/TextSearchFilter.cs:
```diff
@@ -6,6 +6,8 @@
 
 namespace Microsoft.SemanticKernel.Data;
 
+#pragma warning disable CS0618 // FilterClause is obsolete
```
We should decide whether we want to include updating TextSearch to using expressions as part of this update, because if we don't, we shouldn't obsolete the various FilterClause instances yet, since we won't be able to direct developers to an alternative.

roji replied:
Yeah, good point.

I don't think we should tackle updating ITextSearch in this PR, because it's already huge and we probably need to think a bit about the ITextSearch design around this.

I'll go ahead an un-obsolete FilterClause & co. as you suggest - VectorSearchOptions.Filter will remain obsolete, which is the important bit here.

westey-m replied:
Sounds good

adamsitnik commented on dotnet/src/IntegrationTests/Connectors/Memory/Sqlite/SqliteVectorStoreRecordCollectionTests.cs:
```diff
@@ -10,6 +10,8 @@
 
 namespace SemanticKernel.IntegrationTests.Connectors.Memory.Sqlite;
 
+#pragma warning disable CS0618 // VectorSearchFilter is obsolete
```
FWIW we could just add this error to `NoWarn` to `Build.Directory.props` for all test projects

roji replied:
Yeah, that seemed a bit too wide maybe, but you're probably right in retrospect. As I've already done the work and we'll be removing the suppressions anyway for release, I guess we can just keep things as-is for now?

westey-m replied:
Since you already did the work, I think it is nice to have the individual suppressions, since it means that any new inadvertent use of it gets caught.

adamsitnik commented on dotnet/src/InternalUtilities/src/Diagnostics/UnreachableException.cs:
```diff
@@ -0,0 +1,50 @@
+
+#if NETSTANDARD2_0
+
+// Polyfill for using UnreachableException with .NET Standard 2.0
```
What is the advantage of using `UnreachableException` over `ArgumentOutOfRange` exception or `InvalidDataException`?

roji replied:
I use UnreachableException in cases where it's like a debug assertion, i.e. expressing to the reader that if the throw line is reached, then there's a bug in the code itself. This is in contrast e.g. to incorrect user input, which may occur in a correct program - for those I use e.g. ArgumentOutOfRangeException.

So the difference is basically about giving more context to the person reading the code, i.e. conveying that the line should never ever be reached.

westey-m commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchVectorStoreRecordCollection.cs:
```diff
@@ -14,7 +15,7 @@
 using Azure.Search.Documents.Models;
 using Microsoft.Extensions.VectorData;
-using VectorData = Microsoft.Extensions.VectorData;
+using Microsoft.SemanticKernel.Connectors.Postgres;
```
Dodgy copy paste?

roji replied:
Yep!

adamsitnik commented on dotnet/src/VectorDataIntegrationTests/AzureAISearchIntegrationTests/Filter/AzureAISearchFilterFixture.cs:
```diff
@@ -0,0 +1,18 @@
+    // Azure AI search only supports lowercase letters, digits or dashes.
+    protected override string StoreName => "filter-tests";
+
+    public override async Task DisposeAsync()
+        => await base.DisposeAsync();
```
What is  the point of overriding this method? Just to be able to hit the breakpoint here?

roji replied:
Nope, just a leftover - will remove.

adamsitnik reviewed (commented):
So far I've reviewed  101 out of 167 files. I am going to finish it tomorrow, but for now I've found only nits and things that are most likely not very important now like minor perf improvements.

westey-m commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchFilterTranslator.cs:
```diff
@@ -0,0 +1,349 @@
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Microsoft.SemanticKernel.Connectors.Postgres;
```
```suggestion
namespace Microsoft.SemanticKernel.Connectors.AzureAISearch;
```

westey-m commented on dotnet/src/IntegrationTests/Connectors/Memory/MongoDB/MongoDBVectorStoreRecordCollectionTests.cs:
```diff
@@ -12,13 +12,15 @@
     // If null, all tests will be enabled
     private const string? SkipReason = "The tests are for manual verification.";
 
-    [Theory(Skip = SkipReason)]
```
Shall we remove the SkipReason from the others too while we are doing it here?  Also OK with leaving this one in and enabling all MongoDB tests separately, as long as we are consistent.

roji replied:
Ah, this wasn't intentional (I was just testing some stuff) - I'll revert.

We can definitely remove the skipping - I'd just rather keep it out of this PR. But what I'd really prefer is to move the tests into the new integration test infra instead, for them to be shared etc.

westey-m commented on dotnet/src/VectorDataIntegrationTests/AzureAISearchIntegrationTests/AzureAISearchIntegrationTests.csproj:
```diff
@@ -0,0 +1,31 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
```
I know we have a goal of running our tests against more than just .net8.0.  Since we are adding these new, we should consider running against .net framework from the start as well.

roji replied:
Makes sense - am adding net472 to all the new integration test projects (that's the minimum TFM supported by xunit).

Note that I'm also adding a suppression for the entire repo for nullability warnings for the netstandard2.0/net472 TFMs, since nullable reference types aren't really supported there and we get spurious warnings. The important thing is that everything gets verified with net8.0.

westey-m commented on dotnet/src/VectorDataIntegrationTests/AzureAISearchIntegrationTests/Filter/AzureAISearchFilterFixture.cs:
```diff
@@ -0,0 +1,18 @@
+    protected override TestStore TestStore => AzureAISearchTestStore.Instance;
+
+    // Azure AI search only supports lowercase letters, digits or dashes.
+    protected override string StoreName => "filter-tests";
```
I have a goal of running AzureAISearch during CI builds against the instance we have in Azure.  To avoid clashes though the collection names need to be somewhat unique.  I appended a sanitized version of Environment.MachineName to my collections for azure ai search to try and achieve some amount of uniqueness, but haven't tested it out yet.

westey-m replied:
Sure, makes sense.

roji commented on dotnet/SK-dotnet.sln:
```diff
@@ -439,6 +441,30 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "sk-chatgpt-azure-function",
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CosmosMongoDBIntegrationTests", "src\VectorDataIntegrationTests\CosmosMongoDBIntegrationTests\CosmosMongoDBIntegrationTests.csproj", "{11DFBF14-6FBA-41F0-B7F3-A288952D6FDB}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "AzureAISearchIntegrationTests", "src\VectorDataIntegrationTests\AzureAISearchIntegrationTests\AzureAISearchIntegrationTests.csproj", "{06181F0F-A375-43AE-B45F-73CBCFC30C14}"
```
> Currently we do have src and samples folders inside root/dotnet. Since src does not cotain all the source code (samples), I believe we could similarly to other .NET repos add just a test folder with all the test projects.

I agree that moving all tests to a separate `test` folder outside of `src` would make sense and follow typical .NET project structure - but I'll leave that larger question out of this PR (/cc @markwallace-microsoft).

> Moreover, using a word IntegrationTests suggests that these are only integration tests. What if I want to add a single class for the unit test? Should I create a dedicated folder for it?

There currently are also unit test projects for each connectors as well (see e.g. SemanticKernel.Connectors.AzureAISearch.UnitTests); I personally think unit tests for these vector database connectors are less valuable (see #10464), and that we should focus our efforts on good, reliable and reusable integration tests.

But regardless, I agree that if we do have unit tests, separating unit and integration tests to different projects should ideally not be needed. I'm assuming the reason for the split is that integration tests are considered flaky (whereas unit tests are always 100% reliable), so the split allows us to e.g. at least run unit tests in CI. But I believe we should work on making the integration tests reliable (it should be achievable), at which point we can just merge and have a single "tests" project for both integration and unit tests. Another possible reason is running speed (unit tests run faster since they don't access a real database), but I don't think that should be a concern here.

I propose we continue in the direction of reusable integration tests as in this PR (#10194); once that's done, if everything works well and reliably, we can raise the question of what to do with the tests in general (merge integration and unit tests, get rid of the unit tests altogether...).

dluc replied:
a couple of extra historical reasons for the split of unit tests vs integration tests:
* folder structure: e.g. we try to organize unit tests following the hierarchy of the classes under tests, while integration tests are organized differently, e.g. by feature, by scenario, etc
* dependencies: when mixing tests in the same project it's easy to fall in the trap of mocking dependencies during a integration test, which (my pov) should never occur
* similarly, when mixing test types, it can be confusing when big test cases contain both unit and functional tests  I've seen many approaches, such as naming conventions, but eventually separating by project seems to be easier also for new comers
* although it can be achieved also with categories, it's also easy to run all the unit tests without having to worry about external depependencies when unit tests are not mixed with other tests

roji commented on dotnet/samples/Concepts/Memory/VectorStoreLangchainInterop/MappingVectorStoreRecordCollection.cs:
```diff
@@ -1,134 +1,138 @@
-        };
-    }
-}
+// TODO: Commented out as part of implementing LINQ-based filtering, since MappingVectorStoreRecordCollection is no longer easy/feasible.
```
Good point, I'll even use `#if DISABLED_FOR_NOW` (which also allows for nesting arbitrary comments inside)

roji commented on dotnet/src/Connectors/Connectors.AzureCosmosDBMongoDB.UnitTests/AzureCosmosDBMongoDBVectorStoreRecordCollectionTests.cs:
```diff
@@ -13,6 +13,7 @@
 using MongoDB.Driver;
 using Moq;
 using Xunit;
+using MEVD = Microsoft.Extensions.VectorData;
```
This one is unfortunate :(

We have a VectorSearchOptions type, but so does the the Mongo client SDK... Our type is generic, so (AFAIK) we can't do a `using` specifically for the VectorSearchOptions type to disambiguate; so instead I aliased the entire namespace. If you have a better suggestion am definitely interested!

BTW this will - to an extent - be annoying for real users of the Mongo connector. Fortunately, if they use target-typed new it's not a problem, since the type name can simply be omitted:

```c#
_ = await collection.VectorizedSearchAsync(vector, new() { /* options */ });
```

... but if they don't use target-typed new, users will have to do the same namespace aliasing done here. @westey-m @markwallace-microsoft this is probably a wrinkle that we should document...

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchFilterTranslator.cs:
```diff
@@ -0,0 +1,349 @@
+            // If any element contains a comma or space, we switch to using pipe as the delimiter.
+            // If any contains a pipe, we throw (for now).
+            var delimiter = ", ";
+            if (elements.Cast<string>().Any(s => s.Contains(' ') || s.Contains(',')))
```
Oh, I somehow thought IndexOfAny was a new API (and we support netstandard2.0). This isn't very hot, but I made the change since there's no reason not to.

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchFilterTranslator.cs:
```diff
@@ -0,0 +1,349 @@
+            this._filter.Append(", '");
+
+            var isFirst = true;
+            foreach (var element in elements.Cast<string>())
```
Thanks, you're right.. This isn't very hot, but I've optimized the code.

If no element contains comma or space (the default delimiter chars), we do a single efficient loop. But if an element does contain a comma or space, we need to start over since we've already appended commas/spaces as delimiters for earlier elements. So the new code does two loops for that case.

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchFilterTranslator.cs:
```diff
@@ -0,0 +1,349 @@
+
+    private static bool TryGetConstant(Expression expression, out object? constantValue)
+    {
+        switch (expression)
+        {
+            case ConstantExpression { Value: var v }:
+                constantValue = v;
+                return true;
+
+            case var _ when TryGetCapturedValue(expression, out var capturedValue):
+                constantValue = capturedValue;
+                return true;
+
+            default:
+                constantValue = null;
+                return false;
+        }
```
Yeah, I'm definitely an extreme switch guy :) Whenever I see multiple disjoint ifs (if/else if/else if) which check on the same thing (expression in this case), I always find it conceptually better to use a switch, as it expresses the intent very clearly. This case is a tiny bit less conclusive because of the method call (TryGetCapturedValue), but conceptually we're still doing different things for different expression types, so I think the code's intent is still more immediately clear (FWIW expression handling code does this kind of thing a *lot*).

So I'd rather keep the switch :)

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSQLVectorStoreCollectionQueryBuilder.cs:
```diff
@@ -147,44 +152,43 @@ public static QueryDefinition BuildSelectQuery(
-        {
-            return null;
-        }
+        var filterClauses = filter.FilterClauses.ToList();
```
Just for fun: wouldn't ToArray() actually be (silghtly) less efficient than ToList() over arbitrary IEnumerables, since the end result must be of the exact correct size, and so an additional final copy is very likely needed? ToList() can return whatever (possibly oversized capacity) List was left at the end of the enumeration, but ToArray() needs to make a copy at that point, no?

Regardless, I believe that it's not a good idea for a public API to accept an IEnumerable parameter, only to materialize it into an array/List (assuming that materialization is conceptually necessary for the API to do its job). IMHO exposing an IEnumerable conveys to the user that the API simply enumerates internally, and doesn't have to hold all elements in memory at the same time... In other words, if ToArray()/ToList() is going to be necessary, better require the user to do it to make the materialization very explicit in their own code...

Finally, yeah, this is the legacy filtering API which will be removed before release :)

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSqlFilterTranslator.cs:
```diff
@@ -0,0 +1,284 @@
+    {
+        this._storagePropertyNames = storagePropertyNames;
+
+        this._parameters.Clear();
```
You're right - this is a leftover from when I considered reusing the same translator instance. Will remove.

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSqlFilterTranslator.cs:
```diff
@@ -0,0 +1,284 @@
+
+    private void TranslateConstant(ConstantExpression constant)
+    {
+        // TODO: Nullable
```
You're right that more work is needed here.

The actual data types that are supported differ from provider to provider - some providers (e.g. Redis IIRC) support only ints and strings, others support a very wide array of types (e.g. PostgreSQL); so I don't think we want this method to be shared, even if it's currently copy-pasted across.

I've made a note to do a follow-up PR here. What's needed is (integration) testing that would exercise each metadata type that's supported by a connector, including querying - at that point I'll come back and clean this up as well.

roji commented on dotnet/src/Connectors/Connectors.Memory.Sqlite/SqliteVectorStoreCollectionCommandBuilder.cs:
```diff
@@ -263,6 +268,21 @@ private static string GetColumnDefinition(SqliteColumn column)
+
+            whereClause += extraWhereFilter;
+
+            foreach (var p in extraParameters!)
```
Just an implicit assumption that if extraWhereFilter isn't null (checked just above), extraParameter isn't null (these are returned together from SqliteFilterTranslator. I'll add an assertion instead to make this more explicit.

westey-m commented on dotnet/samples/Concepts/Memory/VectorStoreLangchainInterop/MappingVectorStoreRecordCollection.cs:
```diff
@@ -1,134 +1,138 @@
-}
+// TODO: Commented out as part of implementing LINQ-based filtering, since MappingVectorStoreRecordCollection is no longer easy/feasible.
+// TODO: The user provides an expression tree accepting a TPublicRecord, but we require an expression tree accepting a TInternalRecord.
+// TODO: This is something that the user must provide, and is quite advanced.
```
Is the plan to get this checked into the feature branch so long, and then address this area after, or should we keep an eye on this during this PR?

roji replied:
@westey-m I don't have a firm opinion... Importantly, this isn't just "let's defer this for a bit" - the moment users start expressing filter in terms of model (and not storage) properties, any sort of transformation/mapping becomes problematic (as we're discussed before)...

I can imagine an API which allows users to map from model properties (referenced in the expression tree) to arbitrary storage properties, which would plug into the filter translators. For flat record that's easy enough, but for hierarchical ones things could get a bit more complicated...

My personal inclination here is to defer these advanced/custom mapping scenarios - as a whole - until after the initial release, and then take the time to add the proper API... I may be wrong, but IMHO these scenarios shouldn't be part of the everyday user scenario, and we should be able to add support later without breaking. But we should probably discuss this and prioritize.

westey-m replied:
I'm concerned that we are abandoning anyone who is already using advanced mapping until we solve this problem, especially if we remove the legacy filtering, which we said we would do before the initial release.  Definitely something we should discuss more.

westey-m replied:
As discussed, let me dig into the use cases for custom mapping more.  If as suspected it's really a workaround for the lack of hierarchical model support, we should not support custom mapping, and add hierarchical model support.

westey-m commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/Filter/FilterFixtureBase.cs:
```diff
@@ -0,0 +1,184 @@
+}
+
+#pragma warning disable CS1819 // Properties should not return arrays
+#pragma warning disable CA1819 // Properties should not return arrays
```
Duplicate pragma disable and restore below.

westey-m commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/Filter/FilterFixtureBase.cs:
```diff
@@ -0,0 +1,184 @@
+
+#pragma warning disable CS1819 // Properties should not return arrays
+#pragma warning disable CA1819 // Properties should not return arrays
+public class FilterRecord<TKey>
```
Also consider testing with a data model that contains attributes in addition to the definition option, to ensure that the attributes are also correctly used to create indexes and enable filtering.

roji replied:
Absolutely. I see these as separate tests specifically for exercising "model building", i.e. that the provider behaves correctly given the schema/model it's given, across both attributes and VectorStoreRecordDefinition. I think that should be a separate test suite - this one concentrates on all the different things you can express within the filters, and the proper translation of those.

westey-m replied:
We should just be careful about assuming too much about how the vendor implemented their connector, considering that these are specification tests for external connectors too.  Connectors could conceivably have different code paths for definitions and attributes.

roji commented:
@westey-m @dmytrostruk @RogerBarreto @adamsitnik I've addressed/responded to all your comments and fixed all the remaining errors/warnings - please feel free to take another look! Otherwise I can also merge it if everyone's OK with how things look.

westey-m commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/VectorDataIntegrationTests.csproj:
```diff
@@ -0,0 +1,21 @@
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <IsTestProject>true</IsTestProject>
```
Is this not just a test library, rather than a test project itself?

roji replied:
True, will change.

adamsitnik commented on dotnet/src/VectorDataIntegrationTests/Directory.Build.props:
```diff
@@ -0,0 +1,15 @@
+<Project>
+  <Import Project="../../Directory.Build.props" />
```
It's great that we have a props file, so we can move all the things that would be common for all test projects here 👍

roji replied:
Yep. There's some opportunity to refactor common stuff, but I didn't want to go too far in this already-huge PR.

adamsitnik commented on dotnet/src/VectorDataIntegrationTests/InMemoryIntegrationTests/InMemoryIntegrationTests.csproj:
```diff
@@ -0,0 +1,26 @@
+    <RootNamespace>InMemoryIntegrationTests</RootNamespace>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="xunit"/>
+    <PackageReference Include="xunit.runner.visualstudio">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="Microsoft.NET.Test.Sdk"/>
+  </ItemGroup>
```
In one of the future refactorings, we could move this part to the `Directory.Build.props` file

adamsitnik commented on dotnet/src/VectorDataIntegrationTests/MongoDBIntegrationTests/Support/MongoDBTestStore.cs:
```diff
@@ -0,0 +1,51 @@
+
+        this._client = new MongoClient(new MongoClientSettings
+        {
+            Server = new MongoServerAddress(this._container.Hostname, this._container.GetMappedPublicPort(MongoDbBuilder.MongoDbPort)),
```
This is very simple, yet powerful. I am glad to see `Testcontainers` being used, so we no longer need to worry about configuring docker in yaml/json and excluding the test credentials from various tools that search for leaked credentials. 🥇

roji replied:
Oh yes. One additional great advantage of testcontainers, is that a testcontainer is also responsible for waiting until the service inside is ready - the precise details of how to do this are very service-specific, and that's encapsulated in the testcontainer package as well.

adamsitnik commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/Filter/BasicFilterTestsBase.cs:
```diff
@@ -0,0 +1,283 @@
+        bool expectZeroResults = false,
+        bool expectAllResults = false)
+    {
+        var expected = fixture.TestData.AsQueryable().Where(filter).OrderBy(r => r.Key).ToList();
```
This is very clever (the tests don't need to specify the expected results which makes them very simple and we are comparing the results against very reliable LINQ implementation)! 🥇

roji replied:
Well I can't really take the credit for it - most of EF's query tests are done exactly like this ;) Yeah, using LINQ as a querying abstraction is really nice, since you can use the IEnumerable implementation as a free, authoritative oracle.

adamsitnik commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/Filter/FilterFixtureBase.cs:
```diff
@@ -0,0 +1,184 @@
+        await this.Collection.CreateCollectionAsync();
+        await this.SeedAsync();
+
+        // Some databases upsert asynchronously, meaning that our seed data may not be visible immediately to tests.
```
I am glad that you bring your DB experience, so I don't need to spend few days debugging why my tests are flaky ;)

roji replied:
Ahahahaha well actually, most databases I'm used to (relational and to a lesser extent, document databases) don't typically have this problem - it's *usually* important for inserted rows to be immediately visible. Vector databases are a bit special, since querying and updating are very very asymmetrical: they're done by completely different sides of the production system, and typically you don't need immediate visibility of data. So this is a tiny bit new to me too...!

adamsitnik reviewed (approved):
LGTM, great work @roji ! I am very happy to see the `Testcontainers` in action. Also, they way you have implemented the base class for tests is very clever!

roji commented on dotnet/src/VectorDataIntegrationTests/CosmosMongoDBIntegrationTests/Support/CosmosMongoDBTestStore.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    public static CosmosMongoDBTestStore Instance { get; } = new();
+
+    public MongoClient? _client { get; private set; }
+    public IMongoDatabase? _database { get; private set; }
```
Thanks, these are supposed to be private!

dmytrostruk commented:
@roji There is small formatting warning in `WeaviateTestStore.cs` file:
![image](https://github.com/user-attachments/assets/54e1e6ad-be9f-420a-b793-d2414dbeb824)

If I remove it though, it says it can't find `HttpClient` for `net472`:
![image](https://github.com/user-attachments/assets/8b373649-bf10-4485-ad75-92465fe732d8)

For simplicity of this PR, I think we can remove `net472` from target frameworks and handle it in scope of separate task. We also need `net472` for other projects and not only for vector stores.

dmytrostruk commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/VectorDataIntegrationTests.csproj:
```diff
@@ -4,8 +4,7 @@
     <Nullable>enable</Nullable>
-    <IsTestProject>true</IsTestProject>
-    <IsPackable>false</IsPackable>
+    <IsPackable>true</IsPackable>
```
I think this change will force our release pipeline to take this package and release it to NuGet. 
If this project is intended to be published, we should add NuGet-related attributes, like here:
https://github.com/microsoft/semantic-kernel/blob/de2857cfe056aa11f5441ea491c50e7b409e1fab/dotnet/src/Connectors/VectorData.Abstractions/VectorData.Abstractions.csproj#L3-L31

We should also think about proper namespace here, because currently it's just `VectorDataSpecificationTests`.

roji replied:
OK, I'll make IsPackable false for now, as we're not ready to go that far yet. And yeah, we definitely need to finalize package names/namespaces around this.

dmytrostruk commented on dotnet/SK-dotnet.sln:
```diff
@@ -1172,6 +1198,72 @@ Global
+		{27D33AB3-4DFF-48BC-8D76-FB2CDF90B707}.Release|Any CPU.Build.0 = Release|Any CPU
+		{B29A972F-A774-4140-AECF-6B577C476627}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{B29A972F-A774-4140-AECF-6B577C476627}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{B29A972F-A774-4140-AECF-6B577C476627}.Publish|Any CPU.ActiveCfg = Debug|Any CPU
+		{B29A972F-A774-4140-AECF-6B577C476627}.Publish|Any CPU.Build.0 = Debug|Any CPU
```
Small fix for `VectorDataIntegrationTests` project to enable release of this package later:
```suggestion
		{B29A972F-A774-4140-AECF-6B577C476627}.Publish|Any CPU.ActiveCfg = Publish|Any CPU
		{B29A972F-A774-4140-AECF-6B577C476627}.Publish|Any CPU.Build.0 = Publish|Any CPU
```

dmytrostruk commented on dotnet/src/Connectors/Connectors.AzureCosmosDBNoSQL.UnitTests/AzureCosmosDBNoSQLVectorStoreCollectionQueryBuilderTests.cs:
```diff
@@ -181,10 +183,15 @@ public void BuildSearchQueryWithoutFilterDoesNotContainWhereClause()
-            "SELECT x.key,x.property_1,x.property_2 " +
-            "FROM x " +
-            "WHERE (x.key_property = @rk0  AND  x.partition_key_property = @pk0) ";
+        // const string ExpectedQueryText = "" +
+        //     "SELECT x.key,x.property_1,x.property_2 " +
+        //     "FROM x " +
+        //     "WHERE (x.key_property = @rk0  AND  x.partition_key_property = @pk0) ";
```
```suggestion
```

dmytrostruk reviewed (approved):
Small improvements, but in general looks awesome!

westey-m commented on dotnet/samples/Concepts/Memory/VectorStoreLangchainInterop/QdrantFactory.cs:
```diff
@@ -76,6 +76,7 @@ public IVectorStoreRecordCollection<TKey, TRecord> CreateVectorStoreRecordCollec
                 return (collection as IVectorStoreRecordCollection<TKey, TRecord>)!;
             }
 
+#if DISABLED_FOR_NOW // TODO: See note on MappingVectorStoreRecordCollection
```
Should it not be `#if !DISABLED_FOR_NOW`?

roji replied:
Well this basically is just a glorified comment (I did it with #if instead of /* */ mostly because the latter doesn't nest). If I negate, we'd have to define DISABLED_FOR_NOW in the csproj for this to be commented out...

westey-m commented on dotnet/samples/Concepts/Memory/VectorStoreLangchainInterop/MappingVectorStoreRecordCollection.cs:
```diff
@@ -1,5 +1,11 @@
+// TODO: The user provides an expression tree accepting a TPublicRecord, but we require an expression tree accepting a TInternalRecord.
+// TODO: This is something that the user must provide, and is quite advanced.
+
+#if DISABLED
```
Should it be `#if !DISABLED`

roji replied:
Same as above

dluc commented:
> Notably missing at the moment is support for VectorStoreGenericDataModel in LINQ querying; I've opened https://github.com/microsoft/semantic-kernel/issues/10468 and will do this a bit later, with additional planned work on the generic data model.

I could not understand the implications, where/when is `VectorStoreGenericDataModel` used? perhaps an old abstraction that could be deprecated?

dluc commented:
I noticed a comment and discussion about this in the PR: if I understand correctly LINQ filters are expressed using internal storage field names, could you confirm? 

If needed, is there a possibility for mapping external names to internal names? For example, let's assume I build a solution that works with TAGS, and they are stored in different ways, e.g. 
- a list of strings in Postgres, field name "tags"
- a list of string fields in Redis, field names "tag1" "tag2" "tag3" ...
- a JSON property in MongoDb, field name "doc.props.tags"
In this scenario would I be able to transform the incoming LINQ filter on "tags" to another LINQ filter targeting my storage of choice?

dluc commented on dotnet/src/Connectors/Connectors.Memory.Postgres/IPostgresVectorStoreCollectionSqlBuilder.cs:
```diff
@@ -124,13 +126,16 @@ internal interface IPostgresVectorStoreCollectionSqlBuilder
     /// <param name="vectorValue">The vector to match.</param>
-    /// <param name="filter">The filter conditions for the query.</param>
+    /// <param name="legacyFilter">The filter conditions for the query.</param>
+    /// <param name="newFilter">The filter conditions for the query.</param>
```
Renaming parameters is a notable breaking change. Would it make sense to tackle all the breaking changes at once, including removing the old filter and updating the type, to streamline the transition?

roji replied:
This is something we discussed quite extensively recently... The decision was made to obsolete but preserve the current search mechanism, to prevent breakage for current users. We'd remove the legacy search API before releasing, and rename NewFilter to Filter at that point (this creates more churn for users keeping up to date and using the new API).

dluc commented on dotnet/src/Connectors/Connectors.Memory.Postgres/PostgresVectorStoreCollectionSqlBuilder.cs:
```diff
@@ -367,14 +376,24 @@ public PostgresSqlCommandInfo BuildGetNearestMatchCommand(
+#pragma warning disable CS0618 // VectorSearchFilter is obsolete
+        var (where, parameters) = (oldFilter: legacyFilter, newFilter) switch
+        {
+            (not null, not null) => throw new ArgumentException("Either Filter or NewFilter can be specified, but not both"),
```
nit: use nameof(Filter) and nameof(NewFilter) in case params are renamed

roji replied:
Good suggestion - I'll pass in this specific case only because Filter will be removed soon (before release) anyway and this whole piece of code will go away.

dluc commented on dotnet/src/Connectors/Connectors.Memory.Postgres/PostgresFilterTranslator.cs:
```diff
@@ -0,0 +1,332 @@
+
+    private readonly StringBuilder _sql = new();
+
+    internal (string Clause, List<object> Parameters) Translate(
```
do you think there's room for SQL injection, consdiering that the result is injected into a SQL statement?

roji replied:
Absolutely (and that's very important). There's [a test](https://github.com/microsoft/semantic-kernel/pull/10273/files#diff-54e49edd3099d4031761dc5017c0e6bf491e6b334de33f300a54fa1ac3269ee6R24) exercising an input with a single quote to check exactly for this. I have a note to do a general comprehensive pass to make sure everything is airtight - stuff like property names isn't currently safe (those generally aren't user input, but I still need to make them airtight).

dluc commented:
Could you confirm that _NOT_ works also with Contains? example

* NOT Contains over collection field `(r => !r.Tags.Contains("foo"))`

thanks!

westey-m commented on dotnet/src/VectorDataIntegrationTests/RedisIntegrationTests/Support/RedisTestStore.cs:
```diff
@@ -0,0 +1,43 @@
+        await this._container.StartAsync();
+        var redis = await ConnectionMultiplexer.ConnectAsync($"{this._container.Hostname}:{this._container.GetMappedPublicPort(6379)},connectTimeout=60000,connectRetry=5");
+        this._database = redis.GetDatabase();
+        this._defaultVectorStore = new(this._database);
```
We should make sure to test both instances of redis: Json and Hashset.  It's possible to choose the flavour by setting the StorageType option on RedisVectorStoreOptions.  They do differ in some non-obvious ways, especially with regards to how you specify property names.

roji replied:
Totally missed that, thanks... I pushed tests for both. Depending on the exact differences, it might make sense to just treat these as two databases test-wise and have two projects (but for now fixtures and tests are duplicated in the same projects).

roji commented on dotnet/src/SemanticKernel.Core/SemanticKernel.Core.csproj:
```diff
@@ -8,6 +8,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
     <NoWarn>$(NoWarn);SKEXP0001,SKEXP0120</NoWarn>
+<!--    <NoWarn>$(NoWarn);CS1685</NoWarn> &lt;!&ndash; The predefined type 'RuntimeHelpers' is defined in multiple assemblies in the global alias &ndash;&gt;-->
```
Yep, sorry, just a leftover.

roji commented:
@dluc 

>> Notably missing at the moment is support for VectorStoreGenericDataModel in LINQ querying; I've opened https://github.com/microsoft/semantic-kernel/issues/10468 and will do this a bit later, with additional planned work on the generic data model.
>
> I could not understand the implications, where/when is VectorStoreGenericDataModel used? perhaps an old abstraction that could be deprecated?

[VectorStoreGenericDataModel](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/VectorData.Abstractions/VectorStoreGenericDataModel.cs) is basically a way to use this abstraction without a strongly-typed .NET type (e.g. for dynamic scenarios). This is already supported in MEVD, but this PR doesn't add support for it in the new filtering mechanism. This is absolutely something I intend to do.

> I noticed a comment and discussion about this in the PR: if I understand correctly LINQ filters are expressed using internal storage field names, could you confirm?

No - the LINQ filter expression references properties on the user's .NET type, and so references regular .NET properties.

For example, in `collection.VectorizedSearchAsync(vector, new() { NewFilter = r => r.Age > 8 })`, `Age` is a .NET property on a real .NET type (represented here by the lambda parameter `r`). MEVD allows mapping .NET properties to arbitrary storage names (the names that you actually see in the database), and that's fully compatible with the new LINQ filtering (the translator goes through an internal translator table from user-facing property names to storage properties).

> If needed, is there a possibility for mapping external names to internal names? For example, let's assume I build a solution that works with TAGS, and they are stored in different ways, e.g.
>
> a list of strings in Postgres, field name "tags"
> a list of string fields in Redis, field names "tag1" "tag2" "tag3" ...
> a JSON property in MongoDb, field name "doc.props.tags"
In this scenario would I be able to transform the incoming LINQ filter on "tags" to another LINQ filter targeting my storage of choice?

For all of these scenarios, your .NET type would simply have a regular `List<string>` property called `Tags`; referencing it in the LINQ filter and e.g. calling Contains on that property should work simply work on all these databases. One note though: AFAIK in Redis, this would not be mapped to **multiple** string fields, but rather with a single [Tag field](https://redis.io/docs/latest/develop/interact/search-and-query/advanced-concepts/tags/); this already fully works with the Redis connector. This scenario is exercised by [this test](https://github.com/microsoft/semantic-kernel/pull/10273/files#diff-54e49edd3099d4031761dc5017c0e6bf491e6b334de33f300a54fa1ac3269ee6R150).

> Could you confirm that NOT works also with Contains? example

Yes - it does in general. Specific database support varies sometimes (e.g. MongoDB has some general issues with NOT in vector search pre-filters), but it works fine on most databases.

roji commented:
BTW @dluc we're definitely interested in feedback about any other specific filtering features you might be interested, don't hesitate to let us know.

dluc commented:
@roji this is a scenario I'm about to test. 

I'm curious to see this in action since I don't have classes modeling data in storage, every definition is set at runtime. I guess internal field names and external field names will match in every scenario, thus it should just work™️

> > I noticed a comment and discussion about this in the PR: if I understand correctly LINQ filters are expressed using internal storage field names, could you confirm?
>
> No - the LINQ filter expression references properties on the user's .NET type, and so references regular .NET properties.

roji commented:
> I'm curious to see this in action since I don't have classes modeling data in storage, every definition is set at runtime. I guess internal field names and external field names will match in every scenario, thus it should just work

@dluc if I understand correctly, does that mean you using VectorStoreGenericDataModel? If so, note that the LINQ-based filteirng hasn't yet been implemented for that - #10468 tracks that. Full support for dynamic scenarios is absolutely high our list, but some fixes probably need to happen first (see #10802).

So long story short, for dynamic scenarios you'll likely need to wait a bit longer until everything is fully in place.
